### PR TITLE
test: MET-2227 add automation test for transaction overview

### PR DESCRIPTION
**Pull request by tungnguyendinh**
## Description

Please include a summary of the changes and a brief description about this PR

### Issue ticket number and link

- [x] This PR has a valid ticket number or issue: [MET-2227]
- https://cardanofoundation.atlassian.net/browse/MET-2227

### Testing & Validation

- [ ] This PR has been tested/validated in Chrome, Firefox, Safari and Brave browsers.
- [ ] The code has been tested locally with test coverage match expectations.
- [ ] Added new Unit/Component testing (if relevant).

### Security

- [ ] No secrets are being committed (i.e. credentials, PII)
- [ ] This PR does not have any significant security implications

### Code Review

- [ ] There is no unused functionality or blocks of commented out code (otherwise, please explain below)
- [ ] In addition to this PR, all relevant documentation (e.g. Confluence / README.md file) and architecture diagrams (e.g. Miro) were updated

### Design Review

- [ ] If this PR contains changes to the UI, it has gone through a design review with UX Designer or Product owner.
- [ ] In case PR contains changes to the UI, add some screenshots to notice the differences
---
##### _Before_

[comment]: <> (Add screenshots)

##### _After_

[comment]: <> (Add screenshots)

##### _Before_

[comment]: <> (Add screenshots)

##### _After_
image

[comment]: <> (Add screenshots)

##### _Before_

[comment]: <> (Add screenshots)

##### _After_

[comment]: <> (Add screenshots)

[MET-2227]: https://cardanofoundation.atlassian.net/browse/MET-2227?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: playwright/pages/transaction-detail.page.ts [hunk 1]
```diff
@@ -6,8 +6,12 @@ export function transactionDetailPage(page: Page) {
...
+    expect(url, "Check url transaction detail").toContain(`${transactionHash}`);
   };
 
-  return { checkTransactionsDetail };
+  const checkAddressDetail = async ({ addressId }: { addressId: string | null }) => {
+    const url = await page.url();
+    expect(url, "Check url address detail").toContain(`${addressId}`);
+  };
+  return { checkTransactionsDetail, checkAddressDetail };
 }
```
Comment (Sotatek-TaiTruong) [inline playwright/pages/transaction-detail.page.ts:17]: đưa hàm checkAddressDetail vào file address.page.ts

File: playwright/pages/transaction-overview.page.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  const searchBarOnTransaction = async () => {
+    await expect(searchBarEpoch).toHaveText("Transactions");
+  };
+
+  const goToEpochsFromSidebar = async () => {
+    await sidebarBlockchainButton.click();
+    await epochsTab.click();
+  };
+
```
Comment (Sotatek-TaiTruong) [inline playwright/pages/transaction-overview.page.ts:44]: dùng lại hàm trong epoch.page

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+let addressId: string | null;
+let transactionHash: string | null;
+
+Given(/^the user is in the general dashboard page in explorer portal test$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToDashboard();
+});
+When(/^the user selects the Transactions option inside the Blockchain drop down menu$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToEpochsFromSidebar();
+});
+Then(
```
Comment (Sotatek-TaiTruong) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:20]: click vào Transaction ở tab sao lại là goToEpochsFromSidebar

**Review by Sotatek-TaiTruong — APPROVED** [review]

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+When(
+  /^the user selects any data of the row execpt the transaction hash, block number, epoch number or addresses hash of one of the transactions record in the table$/,
+  async ({ page }) => {
+    await transactionOverviewPage(page).openWidget();
+  }
+);
+Then(
+  /^the user should see the Info widget data with the respective data of the selected transaction$/,
+  async ({ page, request }) => {
+    const trxHash = await page.getByTestId("transaction.table.value.txhash#0");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:41]: Steps methods should not have interactions over the pages locators, I understand here you are getting the tx hash from the opened widget and then use that Tx hash for the Koios request. 

A better way to handle this is in the previous steps the `when` step where you open the widget of a transaction, in that moment you capture and return the Tx hash that you will use in the `Then` step. 

Keep in mind that interact with locators is just an action of the Pages classes and the Steps just call functions that has the logic with this locators

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  }
+);
+
+Given(/^the user is in the Transactions page$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToTransactions();
+});
+When(
+  /^the user selects the transaction hash of one record of the transactions table in the transactions page$/,
+  async ({ page }) => {
+    const transactionsTableValueEpoch = page.getByTestId("transaction.table.value.txhash#0");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:57]: Once again is better keep the logic of locators interactions in the respective pages, this code should be encapsulated in a method inside transactionsOverviewPage like open transaction details page by Tx hash or something like that and then you just call that method from here, if you need the value of the TxHash that method should return it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  await transactionDetailPage(page).checkTransactionsDetail({ transactionHash: transactionHash || "" });
+});
+
+Given(/^the user is in the Transactions page portal$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToTransactions();
+});
+When(
+  /^the user selects the block number of one record of the transactions table in the transactions page$/,
+  async ({ page }) => {
+    const transactionsTableValueEpoch = page.getByTestId("transaction.table.value.block#0");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:72]: Same as above comment, let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    await transactionsTableValueEpoch.click();
+  }
+);
+Then(/^the user should be redirected to the block detail page of the selected block number$/, async ({ page }) => {
+  const deatailPageTitle = page.getByTestId("detail.page.title");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:77]: Same as above comment and in this case if the validation is over the block details page that page class should have a method that validate this, in that way maybe for future test we can reuse that method.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+  await expect(deatailPageTitle, "Check title on block detail").toHaveText("Block Details");
+});
+
+Given(/^the user is in the Transactions page web$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToTransactions();
+});
+When(
+  /^the user selects the epoch number of one record of the transactions table in the transactions page$/,
+  async ({ page }) => {
+    const transactionsTableValueEpoch = page.getByTestId("transactions.table.value.epoch#0");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:87]: let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    await transactionsTableValueEpoch.click();
+  }
+);
+Then(/^the user should be redirected to the epoch detail page of the selected epoch number$/, async ({ page }) => {
+  const deatailPageTitle = page.getByTestId("detail.page.title");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:92]: This validation should be executed in the epoch detail page class and here just call the method that executes the validation.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  await expect(deatailPageTitle, "Check title on epoch detail").toHaveText("Epoch Details");
+});
+
+Given(/^the user is in the Transactions page for go to address detail page$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToTransactions();
+});
+When(
+  /^the user selects the input or output address of one record of the transactions table in the transactions page$/,
+  async ({ page }) => {
+    const transactionsTableValueEpoch = page.getByTestId("transaction.table.value.inputAddress#0").first();
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:102]: let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    addressId = await transactionsTableValueEpoch.getAttribute("href");
+    await transactionsTableValueEpoch.click();
+  }
+);
+Then(/^the user should be redirected to the address details page of the selected address hash$/, async ({ page }) => {
+  await transactionDetailPage(page).checkAddressDetail({ addressId: addressId || "" });
+});
+
+Then(/^the user should be redirected to the address detail page of the selected address hash$/, async ({ page }) => {
+  const deatailPageTitle = page.getByTestId("detail.page.title");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:112]: This validation should be executed in the epoch detail page class and here just call the method that executes the validation.

File: playwright/tests/e2e/features/transaction-overview.feature [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    Then the user should be redirected to the epoch detail page of the selected epoch number
+
+  Scenario: Address detail view page through input/output address  
+    Given the user is in the Transactions page for go to address detail page
+    When the user selects the input or output address of one record of the transactions table in the transactions page
+    Then the user should be redirected to the address details page of the selected address hash
+
+  Scenario: Transaction detail page through info widget 
+    Given the user is in the Transactions page for go to transaction detail from detail button in widget
+    Given the user open the info widget of one transaction record in the transactions table of transaction page
```
Comment (Luisdzc) [inline playwright/tests/e2e/features/transaction-overview.feature:35]: You should use And expression avoid double Given expression

File: playwright/tests/e2e/features/transaction-overview.feature [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    When the user selects the view details button in the transcation info widget
+    Then the transaction details page of the selected transaction should be opened
+
+  Scenario: Address detail view page through info widget 
+    Given the user is in the Transactions page to address detail from detail button in widget
+    Given the user open the info widget of one transaction record in transaction web
```
Comment (Luisdzc) [inline playwright/tests/e2e/features/transaction-overview.feature:41]: Again avoid repeat `Given` expression and use `And` expression, besides that this step execute the same action that the step of the previous test where you basically open the widget of a transaction, please keep in mind this I know that maybe in the document I wrote it a bit different but at the end is the same action so we can reuse steps and avoid duplicated code.

File: playwright/tests/e2e/features/transaction-overview.feature [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    When the user selects the Input or Output address hash in the transcation info widget
+    Then the user should be redirected to the address detail page of the selected input address hash
+
+   Scenario: Info widget summary link section
+    Given the user is in the Transactions page for go to transaction detail from summary in widget
+    Given the user opens the info widget of any transaction in the transactions table from widget of transaction
```
Comment (Luisdzc) [inline playwright/tests/e2e/features/transaction-overview.feature:47]: Use `And` expression

File: playwright/tests/e2e/features/transaction-overview.feature [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    When the user selects the summary section in the info widget
+    Then the transaction detail page of the selected transaction should be opened with the summary section displayed    
+
+  Scenario:  Info widget UTXOs link section
+    Given the user is in the Transactions page for go to transaction detail from widget
+    Given the user opens the info widget of any transaction in the transactions table from widget
```
Comment (Luisdzc) [inline playwright/tests/e2e/features/transaction-overview.feature:53]: Use And expression

File: playwright/tests/e2e/features/transaction-overview.feature [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    When the user selects the UTXOs section in the info widget of the selected transaction
+    Then the transaction detail page of the selected transaction should be opened with the UTXOs section displayed   
+
+  Scenario:  Info widget transaction signatores link section
+    Given the user is in the Transactions page in the explorer page for go to transaction table from widget
+    Given the user opens the info widget of any transaction in the transactions table from widget tab
```
Comment (Luisdzc) [inline playwright/tests/e2e/features/transaction-overview.feature:59]: Use And expression

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+  }
+);
+Given(
+  /^the user open the info widget of one transaction record in the transactions table of transaction page$/,
+  async ({ page }) => {
+    await transactionOverviewPage(page).openLastestTransactionWidget();
+  }
+);
+When(/^the user selects the view details button in the transcation info widget$/, async ({ page }) => {
+  const transactionDetailFromWidgetDetailViewBtn = page.getByTestId("transaction.detailViewEpoch.viewDetail");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:129]: let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+});
+
+Given(/^the user is in the Transactions page to address detail from detail button in widget$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToTransactions();
+});
+Given(/^the user open the info widget of one transaction record in transaction web$/, async ({ page }) => {
+  await transactionOverviewPage(page).openLastestTransactionWidget();
+});
+When(/^the user selects the Input or Output address hash in the transcation info widget$/, async ({ page }) => {
+  const addressDetailProducer = page.getByTestId("transaction.widget.inputAddress");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:144]: let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+  }
+);
+Given(
+  /^the user opens the info widget of any transaction in the transactions table from widget of transaction$/,
+  async ({ page }) => {
+    await transactionOverviewPage(page).openLastestTransactionWidget();
+  }
+);
+When(/^the user selects the summary section in the info widget$/, async ({ page }) => {
+  const transactionDetailFromWidgetBySumary = page.getByTestId("transaction.widget.summary");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:168]: let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+);
+
+Given(/^the user is in the Transactions page for go to transaction detail from widget$/, async ({ page }) => {
+  await transactionOverviewPage(page).goToTransactions();
+});
+Given(/^the user opens the info widget of any transaction in the transactions table from widget$/, async ({ page }) => {
+  await transactionOverviewPage(page).openLastestTransactionWidget();
+});
+When(/^the user selects the UTXOs section in the info widget of the selected transaction$/, async ({ page }) => {
+  const transactionDetailFromWidgetByUTXO = page.getByTestId("transaction.widget.utxOs");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:186]: let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

File: playwright/tests/e2e/steps/transactions-overview.steps.ts [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+Given(
+  /^the user opens the info widget of any transaction in the transactions table from widget tab$/,
+  async ({ page }) => {
+    await transactionOverviewPage(page).openLastestTransactionWidget();
+  }
+);
+When(
+  /^the user selects the transaction signatories section in the info widget of the selected transaction$/,
+  async ({ page }) => {
+    const transactionWidgetTrxSignature = page.getByTestId("transaction.widget.signersInformation");
```
Comment (Luisdzc) [inline playwright/tests/e2e/steps/transactions-overview.steps.ts:212]: let's keep locators interactions inside the pages functions and returns the values that you could need in the next steps if need it.

**Review by Luisdzc — CHANGES_REQUESTED** [review]

File: playwright/tests/e2e/features/transaction-overview.feature [hunk 22]
```diff
@@ -32,30 +32,30 @@ Feature: Check Transaction Cardano Explorer
 
   Scenario: Transaction detail page through info widget 
     Given the user is in the Transactions page for go to transaction detail from detail button in widget
-    Given the user open the info widget of one transaction record in the transactions table of transaction page
+    And the user open the info widget of one transaction record in the transactions table of transaction page
```
Comment (Luisdzc) [inline playwright/tests/e2e/features/transaction-overview.feature:35]: Steps in line 35, 41, 47, 53 and 59 are the same step, basically open the info widget of the last transaction we can refactor this step name and re use it in the other test so we don't duplicate code.

**Pull request merged**